functions/strategus-analysis/src/strategus-results/routes.ts

```diff
@@ -214,4 +161,103 @@ export class StrategusResultsRouter {
       }
     );
   }
+
+  private registerUpgradeHandler() {
+    this.server.on("upgrade", async (req, socket, head) => {
+      const url = new URL(req.url ?? "", "http://localhost");
+
+      const match = url.pathname.match(
+        /^\/strategus-results\/([^/]+)\/websocket\/?$/
+      );
+      if (!match) {
+        console.error(
+          `[Strategus Viewer] Invalid upgrade path: ${url.pathname}`
+        );
+        socket.destroy();
+        return;
+      }
+
+      const studyId = match[1];
+      const valid = await validateStudyId(studyId);
+
+      if (!valid) {
+        console.warn(`[Strategus Viewer] Invalid studyId: ${studyId}`);
+        socket.write("HTTP/1.1 404 Not Found\r\n\r\n");
+        socket.destroy();
+        return;
+      }
+
+      const targetUrl = `ws://${encodeURIComponent(studyId)}:3838/websocket`;
+      console.log(
+        `[Strategus Viewer] Upgrade request for studyId: ${studyId} to ${targetUrl}`
+      );
+
+      const strategusWS = new WebSocket(targetUrl);
+
+      strategusWS.on("open", () => {
+        console.log(
+          `[Strategus Viewer] Connected to Strategus WS for study ${studyId}`
+        );
+      });
+
+      strategusWS.on("error", (err) => {
+        console.error(`[Strategus Viewer] Strategus WS error:`, err);
+        if (strategusWS.readyState === WebSocket.OPEN) {
+          strategusWS.close(1011, "Strategus WS error");
+        }
+      });
+
+      // Complete upgrade
+      this.wss.handleUpgrade(req, socket, head, (clientWS) => {
```

Comment:
use `ws` handleupgrade because this function is an express function. using `Deno.upgradeWebSocket`does not work because the request is an express request